packages/pg-gateway/src/auth/md5.ts

```diff
@@ -0,0 +1,151 @@
+import { type BinaryLike, createHash } from 'node:crypto';
+import type { Socket } from 'node:net';
+import type { BufferReader } from 'pg-protocol/dist/buffer-reader';
+import type { Writer } from 'pg-protocol/dist/buffer-writer';
+import type { ConnectionState } from '../connection.types';
+import { BackendMessageCode } from '../message-codes';
+import { BaseAuthFlow } from './base-auth-flow';
+
+export type Md5AuthOptions = {
+  method: 'md5';
+  validateCredentials?: (
+    credentials: {
+      username: string;
+      preHashedPassword: string;
+      salt: Buffer;
+      hashedPassword: string;
+    },
+    connectionState: ConnectionState,
+  ) => boolean | Promise<boolean>;
+  getPreHashedPassword: (
+    credentials: { username: string },
+    connectionState: ConnectionState,
+  ) => string | Promise<string>;
+};
+
+export class Md5AuthFlow extends BaseAuthFlow {
+  private auth: Md5AuthOptions & {
+    validateCredentials: NonNullable<Md5AuthOptions['validateCredentials']>;
+  };
+  private username: string;
+  private salt: Buffer;
+  private completed = false;
+
+  constructor(params: {
+    auth: Md5AuthOptions;
+    username: string;
+    salt?: Buffer;
+    socket: Socket;
+    reader: BufferReader;
+    writer: Writer;
+    connectionState: ConnectionState;
+  }) {
+    super(params);
+    this.auth = {
+      ...params.auth,
+      validateCredentials:
+        params.auth.validateCredentials ??
+        (async ({ preHashedPassword, hashedPassword, salt }) => {
+          const expectedHashedPassword = await hashPreHashedPassword(
+            preHashedPassword,
+            salt,
+          );
+          return hashedPassword === expectedHashedPassword;
+        }),
+    };
+    this.username = params.username;
+    this.salt = params.salt ?? generateMd5Salt();
```

Comment:
This is an oversight on my part, in the spec the salt is generating as part of the auth flow lifecycle, only as a challenge to avoid man-in-the-middle attacks.
So we shouldn't have the possibility to pass the salt as a param here.

> Even better, can we just allow the user to store/send the full hash (expectedHashPassword) instead of us managing the salt separately? Will probably be more secure if they store the full hash with the salt than just the md5(`${user}:${password}`)

On this I would follow what Postgres is doing which will be more predictable.

If the user wants security there is scram now. 😄

Reply:
Sounds good and agreed 👍